shards/shards.go

```diff
@@ -667,13 +667,71 @@ search:
 			r.Priority = r.priority
 			r.MaxPendingPriority = pending.max()
 
-			sender.Send(r.SearchResult)
+			// For simple shards, send the results immediately.
+			if len(r.SearchResult.RepoURLs) <= 1 {
+				sender.Send(r.SearchResult)
+				break
+			}
+
+			// For compound shards, stream results per repo.
+			stats, srs := splitByRepository(r.SearchResult)
+			tr.LazyPrintf("compound shard: results from %d repositories", len(srs))
+			for _, sr := range srs {
+				if len(sr.Files) == 0 {
+					panic("unreachable")
+				}
+				sr.Priority = sr.Files[0].RepositoryPriority
+
+				// TODO (stefan): make sure repos in compound shards are searched in order of their priority.
+				sr.MaxPendingPriority = r.MaxPendingPriority
+				sender.Send(sr)
+			}
+			sender.Send(&zoekt.SearchResult{Stats: stats})
 		}
 	}
 
 	return func() { runtime.KeepAlive(shards) }, err
 }
 
+// splitByRepository splits a zoekt.SearchResult by repository ID. We separate
+// the stats from the searchResult, because stats must be aggregateable.
+func splitByRepository(result *zoekt.SearchResult) (zoekt.Stats, []*zoekt.SearchResult) {
+	if len(result.Files) == 0 {
+		return result.Stats, nil
+	}
+	if len(result.Files) == 1 {
+		stats := result.Stats
+		result.Stats = zoekt.Stats{}
+		return stats, []*zoekt.SearchResult{result}
+	}
+
+	res := []*zoekt.SearchResult{}
```

Comment:
```suggestion
	res := make([]*zoekt.SearchResult, 0, len(result.RepoURLs))
```

Reply:
I moved the sender into `sendByRepository`, hence we don't need the slice anymore.